File: 0337-house-robber-iii/0337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        
        def helper(root: Optional[TreeNode]) -> List[int]:
            if not root:
                return [0, 0]
            
            with_left, without_left = helper(root.left)
            with_right, without_right = helper(root.right)
            
            return [
                root.val + without_left + without_right,
                max(with_left, without_left) + max(with_right, without_right)
            ]
        
        return max(*helper(root))

File: 0337-house-robber-iii/0337-house-robber-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var rob = function(root) {
    return Math.max(...helper(root))
    
    function helper(root) {
        if (!root) return [0, 0]
        const [wLeft, woLeft] = helper(root.left)
        const [wRight, woRight] = helper(root.right)
        return [
            root.val + woLeft + woRight,
            Math.max(wLeft, woLeft) + Math.max(wRight, woRight)
        ]
    }
};